lib/ClangImporter/ClangImporter.cpp

```diff
@@ -7597,9 +7596,53 @@ bool importer::hasEscapableAttr(const clang::RecordDecl *decl) {
   return hasSwiftAttribute(decl, "Escapable");
 }
 
+static const llvm::StringMap<std::vector<int>> STLConditionalEscapableParams{
+    {"vector", {0}},
+    {"array", {0}},
+    {"inplace_vector", {0}},
+    {"deque", {0}},
+    {"forward_list", {0}},
+    {"list", {0}},
+    {"set", {0}},
+    {"flat_set", {0}},
+    {"unordered_set", {0}},
+    {"multiset", {0}},
+    {"flat_multiset", {0}},
+    {"unordered_multiset", {0}},
+    {"stack", {0}},
+    {"queue", {0}},
+    {"priority_queue", {0}},
+    {"tuple", {0}},
+    {"variant", {0}},
+    {"optional", {0}},
+    {"pair", {0, 1}},
+    {"expected", {0, 1}},
+    {"map", {0, 1}},
+    {"flat_map", {0, 1}},
+    {"unordered_map", {0, 1}},
+    {"multimap", {0, 1}},
+    {"flat_multimap", {0, 1}},
+    {"unordered_multimap", {0, 1}},
+    {"span", {0}},   // TODO: remove when span is non-escapable by default
+    {"mdspan", {0}}, // TODO: remove when mdspan is non-escapable by default
+};
+
 std::set<StringRef>
 importer::getConditionalEscapableAttrParams(const clang::RecordDecl *decl) {
   std::set<StringRef> result;
+  if (decl->isInStdNamespace()) {
+    if (const auto *ctsd =
+            dyn_cast<clang::ClassTemplateSpecializationDecl>(decl)) {
+      auto it = STLConditionalEscapableParams.find(ctsd->getName());
+      if (it != STLConditionalEscapableParams.end()) {
+        auto templated = ctsd->getSpecializedTemplate();
+        auto params = templated->getTemplateParameters();
+        for (auto idx : it->second)
+          result.insert(params->getParam(idx)->getName());
```

Comment:
Would this work correctly if the template parameter is unnamed? This probably shouldn't happen with the hand-picked stdlib types (unless Clang picks a redecl of the class template that has an unnamed parameter? not sure if that is possible).

Do you think it would be reasonable to have an assertion here, to make sure there isn't some template weirdness in one of the C++ stdlib versions that we support?

Reply:
Good point! I don't think it would work with unnamed templates. I will add an assertion as it is unlikely that we would run into this with STL implementations.